docs/modules/ROOT/pages/tmail-backend/jmap-extensions/pushWithFirebase.adoc

```diff
@@ -0,0 +1,210 @@
+
+    The server MAY choose to set an expiry if none is given by the client or modify the expiry time given by the client to a shorter duration.
+
+- **types**: `String[]|null`
+A list of types the client is interested in (using the same names as the keys in the *TypeState* object defined in the previous section). A StateChange notification will only be sent if the data for one of these types changes. Other types are omitted from the TypeState object. If `null`, changes will be pushed for all types.
+
+=== FirebaseRegistration/get
+
+Standard */get* method as described in Section 5.1, except it does **not** take or return an *accountId* argument,
+as push subscriptions are not tied to specific accounts. It also does **not** return a *state* argument. The *ids*
+argument may be `null` to fetch all at once.
+
+It does **not** return a *state* argument.
+
+As the *token* properties may contain data that is private to a particular device, the values for these properties MUST NOT be returned.
+
+=== FirebaseRegistration/set
+
+Standard */set* method  as described in Section 5.3, except it does **not** take or return an *accountId* argument,
+as push subscriptions are not tied to specific accounts. It also does **not** take an *ifInState* argument or return
+*oldState* or *newState* arguments.
+
+The client may update the *expires* property to extend (or, less commonly, shorten) the lifetime of a push subscription.
+The server MAY modify the proposed new expiry time to enforce server-defined limits.
+
+Clients SHOULD NOT update or destroy a push subscription that they did not create (i.e., has a *deviceClientId* that
+they do not recognise).
+
+=== Example
+
+At `2018-07-06T02:14:29Z`, a client with deviceClientId `a889-ffea-910` fetches the set of FirebaseRegistration
+currently on the server, making an API request with:
+
+....
+    [[ "FirebaseRegistration/get", {
+      "ids": null
+    }, "0" ]]
+....
+
+Which returns:
+
+....
+    [[ "FirebaseRegistration/get", {
+      "list": [{
+          "id": "e50b2c1d-9553-41a3-b0a7-a7d26b599ee1",
+          "deviceClientId": "b37ff8001ca0",
+          "expires": "2018-07-31T00:13:21Z",
+          "types": [ "Todo" ]
+      }, {
+          "id": "f2d0aab5-e976-4e8b-ad4b-b380a5b987e4",
+          "deviceClientId": "X8980fc",
+          "expires": "2018-07-12T05:55:00Z",
+          "types": [ "Mailbox", "Email", "EmailDelivery" ]
+      }],
+      "notFound": []
+    }, "0" ]]
+....
+
+Since neither of the returned FirebaseRegistration objects have the client's deviceClientId, it knows it does not have a
+current FirebaseRegistration active on the server. So it creates one, sending this request:
+
+....
+    [[ "FirebaseRegistration/set", {
+      "create": {
+        "4f29": {
+          "deviceClientId": "a889-ffea-910",
+          "token": "firebase-token",
```

Comment:
- `token` is different/unique for each deviceClientId or one token (one firebase account) for all devices?

- Does James save this token? Or does it just forward to Firebase?

Reply:
> Does James save this token? 

Of course

> token is different/unique for each deviceClientId or one token (one firebase account) for all devices?

Token is set by customer for each device. Tokens are hidden in /get